### PR TITLE
Changes visitor pass and makes bundle elim implement it.

### PR DESCRIPTION
**Pull request by UnsignedByte**
Adds context info to visitor trait, also resolving #222. Also makes bundle-elim implement visitor after these changes.

### CODE REVIEW

File: src/ir_visitor/visitor.rs [hunk 1]
```diff
@@ -217,27 +260,29 @@ where
         };
 
         // Traverse the commands
-        let mut cmds = std::mem::take(&mut comp.cmds);
-        match self.visit_cmds(&mut cmds, comp) {
+        let mut cmds = std::mem::take(&mut ctx.get_mut(idx).cmds);
```
Comment (rachitnigam) [inline src/ir_visitor/visitor.rs:263]: The way this is currently structured extracts the commands out from the component but gives you mutable access to the context. What I was thinking originally is that we instead completely "detach" the component from the context (by defining a `swap` method on `IndexStore`) and then providing mutable access to the component but immutable access to the context.

Comment (UnsignedByte) [reply src/ir_visitor/visitor.rs:263]: To simplify the number of arguments then could we re-add the "idx()" function on components? BundleElim requires component idx information which currently would have to be passed in separately from everything else.

Comment (rachitnigam) [reply src/ir_visitor/visitor.rs:263]: Oh, I see. Yeah, we can try that but that was causing annoying build problems because we’d need a default idx

Comment (rachitnigam) [reply src/ir_visitor/visitor.rs:263]: One other option is to abstract all of this using a `VisitData` struct which is also more future proof than adding more arguments. Calyx visitor has the same problem where we've slowed made the visitor interface more and more complicated with more arguments

Comment (UnsignedByte) [reply src/ir_visitor/visitor.rs:263]: Personally more of a fan of the second option! I'll work on that next.

File: src/ir_visitor/visitor.rs [hunk 2]
```diff
@@ -29,16 +32,30 @@ impl Action {
     }
 }
 
+/// Contains information passed to visitor functions
+pub struct VisitorData<'comp> {
+    pub ctx: &'comp ir::Context,
+    pub comp: ir::Component,
+    pub idx: ir::CompIdx,
```
Comment (rachitnigam) [inline src/ir_visitor/visitor.rs:39]: Comments for these fields.

File: src/ir_visitor/visitor.rs [hunk 3]
```diff
@@ -29,16 +32,30 @@ impl Action {
     }
 }
 
+/// Contains information passed to visitor functions
+pub struct VisitorData<'comp> {
```
Comment (rachitnigam) [inline src/ir_visitor/visitor.rs:36]: Is the lifetime name correct? I thought the lifetime is that of the entire pass and not just the component?

Comment (rachitnigam) [reply src/ir_visitor/visitor.rs:36]: Oh, I guess you have to change the context after traversing each component. Maybe say that in the comment?

**Review by rachitnigam — APPROVED** [review]
One comment about adding more docs! Merge after addressing and once #255 is done!

**Review by rachitnigam — APPROVED** [review]
LGTM!

**Pull request merged**